### PR TITLE
Implement member type check in interfaces

### PR DESCRIPTION
**Pull request by LemonBoy**
Most of the code is a small refactoring to allow the use of a where_T for signaling the type mismatch, the type checking itself is pretty simple.

### CODE REVIEW

File: src/vim9type.c [hunk 1]
```diff
@@ -744,15 +744,20 @@ type_mismatch_where(type_T *expected, type_T *actual, where_T where)
...
+    if (where.wt_kind == WT_MEMBER)
+    {
+	semsg(_(e_member_str_type_mismatch_expected_str_but_got_str), where.wt_func_name,
+		typename1, typename2);
+    }
+    else if (where.wt_index > 0)
     {
 	if (where.wt_func_name == NULL)
-	    semsg(_(where.wt_variable
+	    semsg(_(where.wt_kind
```
Comment (yegappan) [inline src/vim9type.c]: Is it better to use `where.wt_kind != WT_ARGUMENT` here?
or change the check to `where.wt_kind == WT_ARGUMENT` and reverse the values?

File: src/vim9type.c [hunk 2]
```diff
@@ -744,15 +744,20 @@ type_mismatch_where(type_T *expected, type_T *actual, where_T where)
...
 			 ? e_variable_nr_type_mismatch_expected_str_but_got_str
 		       : e_argument_nr_type_mismatch_expected_str_but_got_str),
 					 where.wt_index, typename1, typename2);
 	else
-	    semsg(_(where.wt_variable
+	    semsg(_(where.wt_kind
```
Comment (yegappan) [inline src/vim9type.c]: Is it better to use `where.wt_kind != WT_ARGUMENT` here?
or change the check to `where.wt_kind == WT_ARGUMENT` and reverse the values?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by chrisbra** [comment]
Thanks. There were some conflicts from #12848 in test_vim9_class.vim  hopefully I resolved them correctly

**Comment by LemonBoy** [comment]
Ah, that was (too) fast! The CI did not kick in for some reason and now master's broken.
This patch should fix the remaining errors, sorry for the noise.

```patch
diff --git a/src/vim9execute.c b/src/vim9execute.c
index 6c208786a..2eb6ba4ca 100644
--- a/src/vim9execute.c
+++ b/src/vim9execute.c
@@ -5451,8 +5451,8 @@ exec_instructions(ectx_T *ectx)
                    {
                        where.wt_index = ct->ct_arg_idx;
                        where.wt_kind = ct->ct_is_var ? WT_VARIABLE : WT_ARGUMENT;
-                       where.wt_func_name = ectx->ec_where.wt_func_name;
                    }
+                   where.wt_func_name = ectx->ec_where.wt_func_name;
                    r = check_typval_type(ct->ct_type, tv, where);
                    if (r == FAIL)
                        goto on_error;
diff --git a/src/vim9type.c b/src/vim9type.c
index 1363a1c3e..4c1693668 100644
--- a/src/vim9type.c
+++ b/src/vim9type.c
@@ -682,8 +682,8 @@ check_typval_arg_type(
     {
        where.wt_index = arg_idx;
        where.wt_kind = WT_ARGUMENT;
-       where.wt_func_name = func_name;
     }
+    where.wt_func_name = func_name;
     return check_typval_type(expected, actual_tv, where);
 }
```

**Comment by chrisbra** [comment]
thanks, applied

**Pull request closed without merging**